src/platforms/virtual/platform.cpp

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License version 2 or 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "platform.h"
+#include "display.h"
+#include "mir/graphics/platform.h"
+#include "options_parsing_helpers.h"
+
+namespace mg = mir::graphics;
+namespace mgv = mir::graphics::virt;
+namespace geom = mir::geometry;
+using namespace std::literals;
+
+
+class mgv::Platform::VirtualDisplayInterfaceProvider : public mg::DisplayInterfaceProvider
+{
+public:
+    explicit VirtualDisplayInterfaceProvider()
+    {
+    }
+
+protected:
+    auto maybe_create_interface(mg::DisplayProvider::Tag const& type_tag)
+        -> std::shared_ptr<mg::DisplayProvider>
+    {
+        class StubGenericEGLDisplayProvider : public GenericEGLDisplayProvider
```

Comment:
This is not actually a stub provider; this is a perfectly sensible `GenericEGLDisplayProvider`, *assuming* that `eglGetDisplay(EGL_DEFAULT_DISPLAY)` returns a valid `EGLDisplay`.